apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.RequestTemplate.cs

```diff
@@ -205,7 +216,27 @@ public sealed class RequestTemplate
             /// an empty collection will be used.</param>
             /// <returns>A new request template.</returns>
             public RequestTemplate WithContentHeaders(IEnumerable<KeyValuePair<string, IEnumerable<string>>> contentHeaders) =>
-                new RequestTemplate(Bucket, ObjectName, HttpMethod, RequestHeaders, contentHeaders.AsReadOnly());
+                new RequestTemplate(Bucket, ObjectName, HttpMethod, RequestHeaders, contentHeaders.AsReadOnly(), QueryParameters);
+
+            /// <summary>
+            /// Returns a new request template with all the same settings as this one
+            /// except for the query parameters.
+            /// </summary>
+            /// <param name="queryParameters">The new query parameters. May be null, in which case
+            /// an empty collection will be used.</param>
+            /// <returns>A new request template.</returns>
+            public RequestTemplate WithQueryParameters(IReadOnlyDictionary<string, IReadOnlyCollection<string>> queryParameters) =>
+                new RequestTemplate(Bucket, ObjectName, HttpMethod, RequestHeaders, ContentHeaders, queryParameters);
+
+            /// <summary>
+            /// Returns a new request template with all the same settings as this one
+            /// except for the query parameters.
+            /// </summary>
+            /// <param name="queryParameters">The new query parameters. May be null, in which case
+            /// an empty collection will be used.</param>
+            /// <returns>A new request template.</returns>
+            public RequestTemplate WithQueryParameters(IEnumerable<KeyValuePair<string, IEnumerable<string>>> queryParameters) =>
+                new RequestTemplate(Bucket, ObjectName, HttpMethod, RequestHeaders, ContentHeaders, queryParameters.AsReadOnly());
```

Comment:
Calling `AsReadOnly()` doesn't make this immutable - nor does the IReadOnlyDictionary-based overload above.
If we want to be able to rely on this not changing under our feet, we should clone it.

That's possibly overkill, especially as these will normally be short-lived objects - it's just worth being aware of.